### PR TITLE
client-go/features: introduce WatchListClient feature gate

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR introduces `WatchListClient` feature gate to client-go.
The gate will be used by `KCM` to enable the streaming feature.
It allows the client to get a stream of individual items instead of chunking from the server.

#### Which issue(s) this PR fixes:

xref: https://github.com/kubernetes/enhancements/issues/3157

#### Special notes for your reviewer:
First we need https://github.com/kubernetes/kubernetes/pull/122555

#### Does this PR introduce a user-facing change?

```release-note
Add `WatchListClient` feature gate to `client-go`. When enabled it allows the client to get a stream of individual items instead of chunking from the server.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md
- [Other doc]: https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs/edit#heading=h.lymszxjovg65
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/features/known_features.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	// // alpha: v1.4
+	// MyFeature featuregate.Feature = "MyFeature"
+	//
+	// Feature gates should be listed in alphabetical, case-sensitive
+	// (upper before any lower case character) order. This reduces the risk
+	// of code conflicts because changes are more likely to be scattered
+	// across the file.
+
+	// owner: @p0lyn0mial
+	// alpha: v1.27
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/features/known_features.go:32]: We didn't have this FG in 1.27 in client-go.
In fact,  you're saying that client-side you're going directly to beta, which I'm fine with [given we gated the logic differently before and it's already there].

But let's not pretend that we had a FG before we actually added it :)

So let's remove it.

@deads2k - FYI

Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/known_features.go]: sgtm, pushed.

File: staging/src/k8s.io/client-go/features/known_features.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	// owner: @p0lyn0mial
+	// beta: v1.30
+	//
+	// Allow the client to get a stream of individual items instead of chunking from the server.
+	//
+	// NOTE:
+	//  The feature is disabled in Beta by default because
+	//  it will only be turned on for selected control plane component(s).
+	WatchList Feature = "WatchListClient"
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/features/known_features.go:39]: Match the featurename please.  WatchListClient

File: staging/src/k8s.io/client-go/features/known_features.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+// defaultKubernetesFeatureGates consists of all known Kubernetes-specific feature keys.
+//
+// To add a new feature, define a key for it above and add it here. The features will be
+// available throughout Kubernetes binaries.
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/features/known_features.go:45]: This comment doesn't appear to be true yet.

Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/features/known_features.go]: sure, updated the comment.

**Pull request merged**